##**[4주차]과제**
- 아래의 과제1), 과제2)의 코드를 완성하시오.
- 모든 코드의 결과를 출력하여, .ipynb의 링크를 **[4주차]/[4주차]과제**에 제출하시오.\
(실습 제출 예시: 4주차_2020XXXX_이름.ipynb 코드 링크)

In [2]:
# 학번, 이름 출력
print(f'학번 : 2353881 이름 : 최대영')

학번 : 12345678 이름 : 홍길동


In [4]:
# google drive 연결
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
# 구글 드라이브에 데이터를 넣고 path 연결 필요
data_path = '/content/drive/MyDrive/과제'

## 과제 1
- 케라스를 이용하여 은닉층이 2개인 MLP를 생성하고 옵티마이저를 Adam, RMSprop, Adadelta, Adagrad로 변경해보면서 성능을 측정한 뒤 어떤 방법이 가장 좋은 성능을 보이는지 확인하고, 각 방법에 대해서 서술하시오.
- 은닉층의 노드 수는 512, 256로 설정
- 은닉층의 Activation Function은 ReLU로 설정
- 모델 레이어 추가  
model.add(tf.keras.layers.Dense(노드갯수, activation='activation_func', kernel_initializer=initializer))  
<출력 예시>  
테스트 정확도 : 0.2779000282287598


In [10]:
# 런타임 재시작 필요
!pip install scikeras==0.13.0 scikit-learn==1.4.2

In [7]:
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
from sklearn.model_selection import GridSearchCV
from scikeras.wrappers import KerasClassifier, KerasRegressor
import os
import numpy as np

In [8]:
# 실험 결과 재현을 위한 seed 고정
SEED = 4
os.environ['PYTHONHASHSEED'] = str(SEED)
os.environ['TF_DETERMINISTIC_OPS'] = '1'
tf.random.set_seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
initializer = tf.keras.initializers.GlorotUniform(seed=SEED)

In [9]:
# 데이터세트 준비
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [11]:
# 데이터 전처리
train_images = train_images.reshape((60000, 784))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 784))
test_images = test_images.astype('float32') / 255

train_labels = tf.keras.utils.to_categorical(train_labels)
test_labels = tf.keras.utils.to_categorical(test_labels)

In [60]:
# 신경망 모델 구축
def build_model():
    network = tf.keras.models.Sequential()

    # 은닉층 생성
    network.add(tf.keras.layers.Dense(512, activation='relu'))
    network.add(tf.keras.layers.Dense(256, activation='relu'))

    network.add(tf.keras.layers.Dense(10, activation='relu'))

    # 옵티마이저 설정 필요
    network.compile(optimizer='adam',
                   loss='mse',
                   metrics=['accuracy'])
    return network

In [61]:
# model 생성
model = build_model()

In [62]:
result = model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 13s 12ms/step - accuracy: 0.8210 - loss: 0.0235
Epoch 2/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 12s 13ms/step - accuracy: 0.8876 - loss: 0.0139
Epoch 3/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 13s 14ms/step - accuracy: 0.8939 - loss: 0.0124
Epoch 4/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 12s 13ms/step - accuracy: 0.8968 - loss: 0.0116
Epoch 5/5
938/938 ━━━━━━━━━━━━━━━━━━━━ 11s 12ms/step - accuracy: 0.8974 - loss: 0.0113


In [63]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8888 - loss: 0.0129


In [64]:
print('테스트 정확도 : ', test_acc)

테스트 정확도 :  0.8888000249862671


### 각 옵티마이저의 성능 작성
- RMSprop : 0.9812999963760376
- Adagrad : 0.5059000253677368
- Adadelta : 0.553600013256073
- Adam : 0.8888000249862671

### 가장 좋은 성능의 옵티마이저 작성
- RMSprop

### 옵티마이저 서술
- RMSprop : RMSprop는 Adagrad에 대한 수정판이다.
- Adagrad : 가변 학습률을 사용하는 방법
- Adadelta : 모멘텀을 이용하여 감소하는 학습률 문제를 처리하는 Adagrad의 변형
- Adam : 기본적으로 'RMSprop + 모멘텀'

## 과제 2 MLP 모델의 정확도(Accuracy)를 71이상 달성하시오.  
- 해당 diabetes.csv 데이터는 Pima 인디언의 의료 기록과 각 환자가 5년 이내에 당뇨병 발병 여부가 저장되어 있다.
- 현재 모델은 1 : 당뇨병 양성, 0 : 당뇨병 음성을 분류하는 MLP 모델이다.
- 해당 모델의 하이퍼 파라미터를 수정하여 정확도 71 이상을 달성하여라.
- 하이퍼 파라미터 : Activation Function, optimizer, epochs, batch size 등


In [76]:
!pip install pandas

In [77]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adagrad, Adadelta, RMSprop, Adam
import pandas as pd
import numpy as np
import os

In [95]:
# 데이터 load
dataset = pd.read_csv(os.path.join(data_path, 'diabetes.csv'))

X = dataset.iloc[:, 0:8].values
y = dataset.iloc[:, 8].values

In [96]:
# 수정 x
def configure_tf_determinism():
    # TensorFlow의 병렬성 감소
    tf.config.experimental.enable_op_determinism()

In [97]:
# 실험 결과 재현을 위한 seed 고정
SEED = 4
os.environ['PYTHONHASHSEED'] = str(SEED)
os.environ['TF_DETERMINISTIC_OPS'] = '1'
tf.random.set_seed(SEED)
np.random.seed(SEED)
tf.random.set_seed(SEED)
initializer = tf.keras.initializers.GlorotUniform(seed=SEED)

In [145]:
# 모델 생성
def build_model():
    model = Sequential()
    # 수정 가능
    model.add(tf.keras.layers.Dense(512, input_dim=8, activation='relu', kernel_initializer=initializer))
    model.add(Dense(10, activation='relu', kernel_initializer=initializer))


    # 수정 x
    model.add(tf.keras.layers.Dense(1, activation='sigmoid', kernel_initializer=initializer))

    model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

    return model

In [146]:
# 수정 x
configure_tf_determinism()

In [147]:
model = build_model()

In [148]:
model.fit(X, y, epochs=100, batch_size=32)

Epoch 1/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.6003 - loss: 0.7789
Epoch 2/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6576 - loss: 0.6537
Epoch 3/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6628 - loss: 0.6424
Epoch 4/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6654 - loss: 0.6320
Epoch 5/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6849 - loss: 0.6203
Epoch 6/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6823 - loss: 0.6142
Epoch 7/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6862 - loss: 0.6096
Epoch 8/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6966 - loss: 0.6021
Epoch 9/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6914 - loss: 0.6011
Epoch 10/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6927 - loss: 0.5946
Epoch 11/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7005 - loss: 0.5891
Epoch 12/100
24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy:

In [149]:
_, accuracy = model.evaluate(X, y)

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7565 - loss: 0.5303  


In [150]:
print('정확도: %.2f' % (accuracy*100))

정확도: 75.65
